# Hands-on Exercise for  FPM Module

### 1. Exploring properties of the dataset accidents_10k.dat. Read more about it here:  http://fimi.uantwerpen.be/data/accidents.pdf

In [1]:
!head accidents_10k.dat

1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 
2 5 7 8 9 10 12 13 14 15 16 17 18 20 22 23 24 25 27 28 29 32 33 34 35 36 37 38 39 
7 10 12 13 14 15 16 17 18 20 25 28 29 30 33 40 41 42 43 44 45 46 47 48 49 50 51 52 
1 5 8 10 12 14 15 16 17 18 19 20 21 22 24 25 26 27 28 29 30 31 41 43 46 48 49 51 52 53 54 55 56 57 58 59 60 61 
5 8 10 12 14 15 16 17 18 21 22 24 25 26 27 28 29 31 33 36 38 39 41 43 46 56 62 63 64 65 66 67 68 
7 8 10 12 17 18 21 23 24 26 27 28 29 30 33 34 35 36 38 41 43 47 59 63 66 69 70 71 72 73 74 75 76 77 78 79 
1 12 14 15 16 17 18 21 22 23 24 25 27 28 29 30 31 35 38 41 43 44 53 56 57 58 59 60 63 66 80 81 82 83 84 
10 12 14 15 16 17 18 21 22 24 25 26 27 28 29 30 31 33 39 41 43 44 46 49 59 60 62 63 66 82 
1 8 10 12 14 15 16 17 18 21 22 23 24 25 27 29 30 31 38 41 43 53 56 59 61 63 66 68 85 86 87 88 89 
1 8 12 13 14 15 16 17 18 22 24 25 28 30 38 41 42 43 46 49 60 63 64 66 80 82 84 90 91 92 93 94 95 


<span style="color:red">**Question 1a:** </span>. How many items are there in the data?

In [2]:
I=!awk -- '{for (i = 1; i <= NF; i++) wc[$i] += 1}; END {print length(wc)}' accidents_10k.dat
Items=int(I[0])
Items

310

310 items

<span style="color:red">**Question 1b:** </span> How many transactions are present in the data?

In [3]:
!wc -l accidents_10k.dat 

10000 accidents_10k.dat


10000 transactions

<span style="color:red">**Question 1c:** </span>.  What is the length of the smallest transaction?

In [4]:
a=!awk '{print NF}' accidents_10k.dat
b=a.sort()
print (int(b[0]))

23


Length of smallest transaction is 23

<span style="color:red">**Question 1d:** </span>  What is the length of the longest transaction?

In [5]:
print (int(b[-1]))

45


Length of longest transaction is 45

<span style="color:red">**Question 1e:** </span>  What is the size of the search space of frequent itemsets in this data?

In [6]:
2**(Items)

2085924839766513752338888384931203236916703635113918720651407820138886450957656787131798913024

<span style="color:green">**Answer:** </span> 

<span style="color:red">**Question 1f:** </span> 
Assume that you work for the deparment of transportation that collected this data. What benefit do you see in using itemset mining approaches on this data?

I can find the most prevailing circumstances for the occurrence of accidents.

<span style="color:red">**Question 1g:** </span>  What type of itemsets (frequent, maximial or closed) would you be interested in discovering this dataset? State your reason.

I prefer closed itemsets. Frequent itemsets for the given data will be very high in number with which we run short of computation power. Maximal sets provide summary of the frequent itemsets but there turns out to be some loss in the support of each subset and we can't state the support of it's subsets whereas closed itemsets are a lossless summary of all possible frequent itemsets.

<span style="color:red">**Question 1h:** </span>  What minsup threshold would you use and why?

It depends upon the number of transactions and items. We can't predict the minsup by just observing the dataset. May be in this case, minsup of 6000 can be taken for better computation(Frequent itemsets are a bit less).


### 2. Generating frequent, maximal and closed itemsets using $\color{red}{\text{Apriori}}$, $\color{red}{\text{ECLAT}}$, and $\color{red}{\text{FPGrowth}}$ algorihtms from the dataset accidents_10k.dat 

<span style="color:red">**Question 2a:** </span> Generate frequent itemsets using Apriori, for minsup = 2000, 3000, and 4000. Which of these minsup thresholds results in a maximum number of frequent itemsets? Which of these minsup thresholds results in a least number of frequent itemsets? Provide a rationale for these observations.

In [7]:
!chmod u+x apriori eclat fpgrowth prefixspan seqwog
!./apriori
!./eclat
!./fpgrowth
!./prefixspan
!./seqwog

usage: ./apriori [options] infile [outfile]
find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
-t#      target type                              (default: s)
         (s: frequent, c: closed, m: maximal item sets,
          g: generators, r: association rules)
-m#      minimum number of items per set/rule     (default: 1)
-n#      maximum number of items per set/rule     (default: no limit)
-s#      minimum support of an item set/rule      (default: 10%)
-S#      maximum support of an item set/rule      (default: 100%)
         (positive: percentage, negative: absolute number)
-o       use original rule support definition     (body & head)
-c#      minimum confidence of an assoc. rule     (default: 80%)
-e#      additional evaluation measure            (default: none)
-a#      aggregation mode for evaluation measure  (default: none)
-d#      threshold for add. evaluation measure    (default: 10%)
-i       invalidate eva

PrefixSpan version 1.00 - Sequential Pattern Miner
Written by Yasuo Tabei


Usage: prefixspan [OPTION]... INFILE

       where [OPTION]...  is a list of zero or more optional arguments
             INFILE(s)    is the name of the input transaction database

Additional arguments (at most one input file may be specified):
       -min_sup [minimum support]
       -max_pat [maximum pattern]

usage: ./seqwog [options] infile [outfile]
find frequent sequences without gaps
version 3.16 (2016.10.15)        (c) 2010-2016   Christian Borgelt
-t#      target type                              (default: s)
         (s: frequent, c: closed, m: maximal sequences, r: rules)
         (target type 'r' implies -a (all occurrences))
-m#      minimum number of items per sequence     (default: 1)
-n#      maximum number of items per sequence     (default: no limit)
-s#      minimum support of a sequence            (default: 10%)
         (positive: percentage, negative: absolute number)
-o       use origina

In [8]:
!./apriori -ts -s-2000 accidents_10k.dat accidents_2000_sup.txt

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.01s].
building transaction tree ... [20250 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 13 done [19.09s].
writing accidents_2000_sup.txt ... [851034 set(s)] done [0.11s].


In [9]:
!./apriori -ts -s-3000 accidents_10k.dat accidents_3000_sup.txt

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.01s].
filtering, sorting and recoding items ... [38 item(s)] done [0.00s].
sorting and reducing transactions ... [9674/10000 transaction(s)] done [0.01s].
building transaction tree ... [24741 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 done [4.31s].
writing accidents_3000_sup.txt ... [133799 set(s)] done [0.02s].


In [10]:
!./apriori -ts -s-4000 accidents_10k.dat accidents_4000_sup.txt

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [33 item(s)] done [0.00s].
sorting and reducing transactions ... [9381/10000 transaction(s)] done [0.00s].
building transaction tree ... [22267 node(s)] done [0.01s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 done [1.33s].
writing accidents_4000_sup.txt ... [29501 set(s)] done [0.00s].


Minsup 4000 gives least number of frequent sets because it checks for only those frequent sets whose support>=4000 and ignores all the others.

<span style="color:red">**Question 2b:** </span>   Using Apriori, compare the execution time for finding frequent itemsets for minsup = 2000, 3000, and 4000. Which of these minsup thresholds takes the least amount of time? Provide a rationale for this observation.

In [11]:
import datetime 
start = datetime.datetime.now()
!./apriori -ts -s-2000 accidents_2000_sup.dat 
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt

./apriori: cannot open file accidents_2000_sup.dat
0 secs  229401 microsecs


In [12]:
import datetime 
start = datetime.datetime.now()
!./apriori -ts -s-3000 accidents_3000_sup.dat 
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt

./apriori: cannot open file accidents_3000_sup.dat
0 secs  238499 microsecs


In [13]:
import datetime 
start = datetime.datetime.now()
!./apriori -ts -s-4000 accidents_4000_sup.dat 
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt

./apriori: cannot open file accidents_4000_sup.dat
0 secs  246645 microsecs


Minsup 4000 takes less time  

<span style="color:red">**Question 2c:** </span> Using Apriori, find the frequent itemsets for minsup = 2000, 3000, and 4000. Determine the number of itemsets for each size (1 to max length of an itemset). What trends do you see that are common for all three minsup thresholds? What trends do you see that are different? Provide a rationale for these observations.

In [14]:
from collections import Counter 
a=!awk '{print NF-1}' accidents_2000_sup.txt
b=list(zip(Counter(a).keys(), Counter(a).values()))
b.sort(key = lambda x: int(x[0]))
b

[('1', 49),
 ('2', 705),
 ('3', 5285),
 ('4', 23745),
 ('5', 69647),
 ('6', 139628),
 ('7', 195730),
 ('8', 193299),
 ('9', 133819),
 ('10', 63937),
 ('11', 20497),
 ('12', 4189),
 ('13', 483),
 ('14', 21)]

In [15]:
!awk '{print NF-1}' accidents_3000_sup.txt|sort -n|uniq -c

     38 1
    468 2
   2830 3
   9887 4
  21779 5
  31964 6
  32020 7
  21862 8
   9839 9
   2705 10
    387 11
     20 12


In [16]:
!awk '{print NF-1}' accidents_4000_sup.txt|sort -n|uniq -c

     33 1
    319 2
   1492 3
   4043 4
   6926 5
   7751 6
   5626 7
   2546 8
    668 9
     91 10
      6 11


Number of itemsets are greatly reduced from minsup 2000 to 4000. 
Similarity : Count of frequent itemsets with length 1 to 6 increase and then it decreases which is similar to the normal distribution.
Difference : Count of frequent itemsets (for every length) decreases from minsup 2000 to minsup 4000.
Rationale : We only check for the itemsets whose support >= 4000 in minsup 4000 case which rules out itemsets.

<span style="color:red">**Question 2d:** </span>  Using Apriori with minsup=2000, compare the number of frequent, maximal, and closed itemsets. Which is the largest set and which is the smallest set? Provide a rationale for these observations.

In [17]:
!./apriori -tm -s-4000 accidents_4000_sup.txt 

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_4000_sup.txt ... [3336 item(s), 29501 transaction(s)] done [0.01s].
filtering, sorting and recoding items ... [15 item(s)] done [0.00s].
sorting and reducing transactions ... [13503/29501 transaction(s)] done [0.00s].
building transaction tree ... [13503 node(s)] done [0.00s].
checking subsets of size 1 2 3 done [0.00s].
filtering for maximal item sets ... done [0.00s].
writing <null> ... [40 set(s)] done [0.00s].


In [18]:
!./apriori -tc -s-4000 accidents_4000_sup.txt

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_4000_sup.txt ... [3336 item(s), 29501 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [15 item(s)] done [0.00s].
sorting and reducing transactions ... [13503/29501 transaction(s)] done [0.00s].
building transaction tree ... [13503 node(s)] done [0.00s].
checking subsets of size 1 2 3 done [0.01s].
filtering for closed item sets ... done [0.00s].
writing <null> ... [50 set(s)] done [0.00s].


In [19]:
!./apriori -ts -s-4000 accidents_10k.dat accidents_4000_sup.txt

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.01s].
filtering, sorting and recoding items ... [33 item(s)] done [0.00s].
sorting and reducing transactions ... [9381/10000 transaction(s)] done [0.00s].
building transaction tree ... [22267 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 done [1.26s].
writing accidents_4000_sup.txt ... [29501 set(s)] done [0.01s].


Largest  : Frequent itemsets(F)
Smallest : Maximal itemsets(M=Itemsets which have no frequent supersets)
Closed itemsets(C) are the ones which have no frequent supersets of same support.
Rationale : 
M ⊆ C ⊆ F

<span style="color:red">**Question 2e:** </span> For a minsup = 2000, compare the execution time for Apriori, ECLAT and FPGrowth. Which of these algorithms took the least amount of time. Provide a rationale for this observation.

In [20]:
import datetime 
start = datetime.datetime.now()
!./apriori -ts -s-2000 accidents_10k.dat
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.01s].
building transaction tree ... [20250 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 13 done [19.09s].
writing <null> ... [851034 set(s)] done [0.02s].
19 secs  691012 microsecs


In [21]:
import datetime 
start = datetime.datetime.now()
!./eclat -ts -s-2000 accidents_10k.dat 
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./eclat - find frequent item sets with the eclat algorithm
version 5.20 (2017.05.30)        (c) 2002-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.01s].
writing <null> ... [851034 set(s)] done [0.28s].
0 secs  557436 microsecs


In [22]:
import datetime 
start = datetime.datetime.now()
!./fpgrowth -ts -s-2000 accidents_10k.dat
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./fpgrowth - find frequent item sets with the fpgrowth algorithm
version 6.17 (2017.05.30)        (c) 2004-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.01s].
writing <null> ... [851034 set(s)] done [0.07s].
0 secs  343202 microsecs


fp Growth perform much better than the other two. 
In Apriori, candidate set generation and Database scans(equal to the max length of the itemset) happens. Database scan is highly expensive incase of large Database.
Whereas in ECLAT, although candidate set generation and Database scan happens. Database is scanned only once because it uses an inverted index to store the dataset which is better compared to Apriori.


<span style="color:red">**Question 2f:** </span> For a minsup = 4000, compare the execution time for Apriori, ECLAT and FPGrowth. Which of these algorithms took the least amount of time. Provide a rationale for this observation.

In [23]:
import datetime 
start = datetime.datetime.now()
!./apriori -ts -s-4000 accidents_10k.dat
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [33 item(s)] done [0.00s].
sorting and reducing transactions ... [9381/10000 transaction(s)] done [0.01s].
building transaction tree ... [22267 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 done [1.27s].
writing <null> ... [29501 set(s)] done [0.01s].
1 secs  567695 microsecs


In [24]:
import datetime 
start = datetime.datetime.now()
!./eclat -ts -s-4000 accidents_10k.dat
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./eclat - find frequent item sets with the eclat algorithm
version 5.20 (2017.05.30)        (c) 2002-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.01s].
filtering, sorting and recoding items ... [33 item(s)] done [0.00s].
sorting and reducing transactions ... [9381/10000 transaction(s)] done [0.02s].
writing <null> ... [29501 set(s)] done [0.04s].
0 secs  288465 microsecs


In [25]:
import datetime 
start = datetime.datetime.now()
!./fpgrowth -ts -s-4000 accidents_10k.dat
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./fpgrowth - find frequent item sets with the fpgrowth algorithm
version 6.17 (2017.05.30)        (c) 2004-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [33 item(s)] done [0.00s].
sorting and reducing transactions ... [9381/10000 transaction(s)] done [0.01s].
writing <null> ... [29501 set(s)] done [0.01s].
0 secs  275427 microsecs


FpGrowth performs better.In Apriori, candidate set generation and Database scans(equal to the max length of the itemset) happens. Database scan is highly expensive incase of large Database.
Whereas in ECLAT, although candidate set generation and Database scan happens. Database is scanned only once because it uses an inverted index to store the dataset which is better compared to Apriori.

<span style="color:red">**Question 2g:** </span>  For a minsup = 6000, compare the execution time for Apriori, ECLAT and FPGrowth. Which of these algorithms took the least amount of time. Provide a rationale for this observation.

In [26]:
import datetime 
start = datetime.datetime.now()
!./apriori -ts -s-6000 accidents_10k.dat
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.01s].
filtering, sorting and recoding items ... [20 item(s)] done [0.00s].
sorting and reducing transactions ... [3216/10000 transaction(s)] done [0.01s].
building transaction tree ... [6478 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 done [0.03s].
writing <null> ... [2254 set(s)] done [0.00s].
0 secs  304428 microsecs


In [27]:
import datetime 
start = datetime.datetime.now()
!./eclat -ts -s-6000 accidents_10k.dat
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./eclat - find frequent item sets with the eclat algorithm
version 5.20 (2017.05.30)        (c) 2002-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.03s].
filtering, sorting and recoding items ... [20 item(s)] done [0.00s].
sorting and reducing transactions ... [3216/10000 transaction(s)] done [0.00s].
writing <null> ... [2254 set(s)] done [0.01s].
0 secs  270384 microsecs


In [28]:
import datetime 
start = datetime.datetime.now()
!./fpgrowth -ts -s-6000 accidents_10k.dat
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./fpgrowth - find frequent item sets with the fpgrowth algorithm
version 6.17 (2017.05.30)        (c) 2004-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [20 item(s)] done [0.00s].
sorting and reducing transactions ... [3216/10000 transaction(s)] done [0.00s].
writing <null> ... [2254 set(s)] done [0.01s].
0 secs  265278 microsecs


FpGrowth performs better.In Apriori, candidate set generation and Database scans(equal to the max length of the itemset) happens. Database scan is highly expensive incase of large Database. Whereas in ECLAT, although candidate set generation and Database scan happens. Database is scanned only once because it uses an inverted index to store the dataset which is better compared to Apriori.

<span style="color:red">**Question 2h:** </span> Fill the following table based on execution times computed in __2e__, __2f__, and __2g__. State your observations on the relative computational efficiency at different support thresholds. Based on your knowledge of these algorithms, provide the reasons behind your observations.

|   Algorithm                |minsup=2000         |minsup=4000         |minsup=6000         |
|----------------------------|--------------------|--------------------|--------------------|    
|Apriori                     |19 secs  691012 ms  |1 secs  567695ms    |          304428    |
|Eclat                       |557436              |288465              |          270384    |
|FPGrowth                    |343202              |275427              |          265278    |

FpGrowth performs better.In Apriori, candidate set generation and Database scans(equal to the max length of the itemset) happens. Database scan is highly expensive incase of large Database. Whereas in ECLAT, although candidate set generation and Database scan happens. Database is scanned only once because it uses an inverted index to store the dataset which is better compared to Apriori.

### 3. Discovering frequent subsequences and substrings

Assume that roads in a Cincinnati are assigned numbers. Participants are enrolled in a transportation study and for every trip they make using their car, the sequence of roads taken are recorded. Trips that involves freeways are excluded. This data is in the file <span style="color:blue">road_seq_data.dat</span>.

<span style="color:red">**Question 3a:** </span>  What 'type' of sequence mining will you perform to determine frequently taken 'paths'? Paths are sequences of roads traveresed consecutively in the same order.

Projection or vertical based

<span style="color:red">**Question 3b:** </span> How many sequences are there in this sequence database?

In [29]:
!wc -l road_seq_data.dat

1000 road_seq_data.dat


1000 sequences

<span style="color:red">**Question 3c:** </span> What is the size of the alphabet in this sequence database?

In [30]:
S=!awk -- '{for (i = 1; i <= NF; i++) wc[$i] += 1}; END {print length(wc)}' road_seq_data.dat
ns=int (S[0])
ns


1283

1283 

<span style="color:red">**Question 3d:** </span> What are the total number of possible subsequences of length 2 in this dataset?

In [31]:
ns*ns

1646089

<span style="color:red">**Question 3e:** </span> What are the total number of possible substrings of length 2 in this dataset?

In [32]:
ns-2+1

1282

<span style="color:red">**Question 3f:** </span> Discover frequent __subsequences__ with minsup = 10 and report the number of subsequences discovered.

In [33]:
!./prefixspan -min_sup 10 road_seq_data.dat

PrefixSpan version 1.00 - Sequential Pattern Miner
Written by Yasuo Tabei


( 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232 233 234 235 236 237 238 239 240 241 242 243 244 245 246 247 248 249 250 251 252 253 254 255 256 257 

( 160 310 456 498 575 577 583 698 821 824 908 922 945 950 974 982 987 ) : 17
66 
( 57 58 95 144 145 221 222 277 298 299 386 408 724 ) : 13
73 
( 9 13 17 21 49 53 136 140 213 217 262 290 294 345 376 390 432 471 490 623 856 866 884 902 909 943 948 972 980 985 ) : 30
73 37605 
( 9 13 17 21 49 53 136 140 213 217 262 290 294 345 376 432 471 623 948 985 ) : 20
74 
( 10 14 18 22 50 55 137 142 214 219 263 291 296 391 ) : 14
74 37613 
( 10 14 18 22 50 137 214 263 291 391 ) : 10
74 37613 37630 
( 10 14 18 22 50 137 214 263 291 391 ) : 10
74 37613 37630 37664 
( 10 14 18 22 50 137 214 263 291 391 ) : 10
74 37613 37630 37664 37673 
( 10 14 18 22 50 137 214 263 291 391 ) : 10
74 37613 37630 37673 
( 10 14 18 22 50 137 214 263 291 391 ) : 10
74 37613 37664 
( 10 14 18 22 50 137 214 263 291 391 ) : 10
74 37613 37664 37673 
( 10 14 18 22 50 137 214 263 291 391 ) : 10
74 37613 37673 
( 10 14 18 22 50 137 214 263 291 391 ) : 10
74 37630 
( 10 14 18 22 50 137 214 263 291 391 ) : 10
74 37630 37664 
( 10 1

74725 74738 74776 79279 37613 
( 163 312 423 439 467 524 566 709 803 833 ) : 10
74725 74738 76435 
( 409 602 624 654 679 717 781 792 806 877 888 992 998 ) : 13
74725 74738 76435 13 
( 409 602 624 654 679 717 781 792 806 877 992 998 ) : 12
74725 74738 77728 
( 28 327 429 628 647 669 738 769 811 825 841 925 932 961 977 ) : 15
74725 74738 77728 13 
( 28 429 628 647 738 769 811 825 841 925 932 961 977 ) : 13
74725 74738 78038 
( 32 101 122 184 259 331 372 416 418 436 441 528 549 589 605 656 672 698 939 966 994 ) : 21
74725 74738 78038 13 
( 32 101 122 184 259 331 372 416 418 436 441 528 549 589 605 656 672 698 939 966 994 ) : 21
74725 74738 78038 13 18 
( 32 101 122 184 259 418 528 605 656 672 939 966 994 ) : 13
74725 74738 78038 18 
( 32 101 122 184 259 418 528 605 656 672 939 966 994 ) : 13
74725 74738 79279 
( 163 237 312 379 413 423 439 445 449 467 500 509 524 539 566 571 632 709 803 833 927 ) : 21
74725 74738 79279 13 
( 237 379 413 439 445 449 467 500 509 524 539 566 571 632 709 803 

74731 74780 13 38 
( 26 65 73 80 87 131 150 168 178 208 227 242 248 254 285 304 321 339 363 388 400 404 435 466 495 523 588 620 653 722 730 735 747 750 759 762 777 794 796 844 858 868 886 904 911 935 953 ) : 47
74731 74780 13 38 231 
( 65 87 131 150 178 208 254 750 762 796 935 ) : 11
74731 74780 13 38 231 37613 
( 65 87 131 150 178 208 254 750 762 796 ) : 10
74731 74780 13 38 37613 
( 65 73 80 87 131 150 168 178 208 227 242 248 254 285 304 321 363 388 400 404 495 523 588 620 653 722 730 750 762 777 796 844 ) : 32
74731 74780 13 38 37613 37638 
( 65 73 80 87 131 150 168 178 208 227 242 248 254 285 304 321 363 388 400 404 495 523 588 653 722 730 844 ) : 27
74731 74780 13 38 37638 
( 65 73 80 87 131 150 168 178 208 227 242 248 254 285 304 321 363 388 400 404 495 523 588 653 722 730 844 ) : 27
74731 74780 13 231 
( 65 87 131 150 178 208 254 750 762 796 935 ) : 11
74731 74780 13 231 37613 
( 65 87 131 150 178 208 254 750 762 796 ) : 10
74731 74780 13 37613 
( 65 73 80 87 131 150 168 178 208

74791 
( 8 12 16 20 48 52 135 139 212 216 261 289 293 344 375 389 431 470 489 622 855 865 883 901 908 942 947 971 979 984 ) : 30
74791 5 
( 8 12 16 20 48 52 135 139 212 216 261 289 293 344 375 431 470 622 947 984 ) : 20
74791 5 37613 
( 8 12 16 20 52 139 216 293 375 622 ) : 10
74791 37613 
( 8 12 16 20 52 139 216 293 375 389 622 855 865 ) : 13
74792 
( 9 13 17 21 49 54 136 141 213 218 262 290 295 390 ) : 14
74792 13 
( 9 13 17 21 49 136 213 262 290 390 ) : 10
74792 13 30 
( 9 13 17 21 49 136 213 262 290 390 ) : 10
74792 13 30 64 
( 9 13 17 21 49 136 213 262 290 390 ) : 10
74792 13 30 64 73 
( 9 13 17 21 49 136 213 262 290 390 ) : 10
74792 13 30 73 
( 9 13 17 21 49 136 213 262 290 390 ) : 10
74792 13 64 
( 9 13 17 21 49 136 213 262 290 390 ) : 10
74792 13 64 73 
( 9 13 17 21 49 136 213 262 290 390 ) : 10
74792 13 73 
( 9 13 17 21 49 136 213 262 290 390 ) : 10
74792 30 
( 9 13 17 21 49 136 213 262 290 390 ) : 10
74792 30 64 
( 9 13 17 21 49 136 213 262 290 390 ) : 10
74792 30 64 73 
( 9 

In [34]:
!./prefixspan -min_sup 10 road_seq_data.dat| sed -n 'p;n'| wc -l

PrefixSpan version 1.00 - Sequential Pattern Miner
Written by Yasuo Tabei

4589


4589 sequences

<span style="color:red">**Question 3g:** </span>  Discover frequent __substrings__ with minsup = 10 and report the number of substrings discovered.

In [35]:
!./seqwog -ts -s-10 road_seq_data.dat road_sub_string

./seqwog - find frequent sequences without gaps
version 3.16 (2016.10.15)        (c) 2010-2016   Christian Borgelt
reading road_seq_data.dat ... [1283 item(s), 1000 transaction(s)] done [0.00s].
recoding items ... [1283 item(s)] done [0.00s].
reducing and triming transactions ... [844/1000 transaction(s)] done [0.01s].
writing road_sub_string ... [613 sequence(s)] done [0.00s].


In [36]:
!wc -l road_sub_string

613 road_sub_string


613 substrings

<span style="color:red">**Question 3h:** </span> Explain the difference in the number of frequent subsequences and substrings found in __3f__ and __3g__ above.

Consecutive sequence of symbols is substring whereas a subsequence need not have consecutive symbols(There can be gaps between symbols). Hence we can have many combination of subsequence when compared to substring.